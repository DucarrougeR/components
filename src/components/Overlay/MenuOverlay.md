```js noeditor
const StatusAndResources = require('../../../styleguide_components/StatusAndResources')
  .StatusAndResources
;<StatusAndResources
  status="stable"
  githubURL="https://github.com/looker/lens/blob/master/src/components/Overlays/MenuOverlay.tsx"
  feedbackTitle="Popover Component Feedback"
/>
```

### Menu Overlay

The `MenuOverlay` component is composed of an `Overlay` with a transparent backdrop. The placement of the backdrop is also customizable.

Try scrolling down so the two menu buttons are at the top of the screen. The backdrop is adjusted 50px from the top, so clicking the other menu button while the current menu overlay is still open will close the current menu and open the new menu overlay.

```js
const Menus = () => {
  const content = (<Box width="260px" maxHeight="90vh" my="none" style={{ overflowY: 'auto' }}>
    <Text>Menu Content</Text>
    <List>
      <ListItem>Thing 1</ListItem>
      <ListItem>Thing 2</ListItem>
    </List>
  </Box>
  )
  return (
  <Flex justifyContent="space-evenly">
    <MenuOverlay placement="bottom" backdrop={{ top: '50px' }} content={content}>
      <Button>Menu One</Button>
    </MenuOverlay>
    <MenuOverlay placement="bottom" backdrop={{ top: '50px' }} content={content}>
      <Button>Menu Two</Button>
    </MenuOverlay>
  </Flex>
  )

}

<Menus />
```
