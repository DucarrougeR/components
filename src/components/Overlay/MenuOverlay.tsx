import * as React from 'react'
import { palette, shadows } from '../../style'
import { radii } from '../../style/radii'
import { CustomizableAttributes } from '../../types/attributes'
import { ManagedModalProps, ModalSurfaceStyleProps } from '../Modal'
import { ModalManager, ModalManagerProps } from '../Modal/ModalManager'
import {
  Overlay,
  OverlayChildrenProps,
  OverlayInteractiveProps,
  OverlaySurface,
} from './'

export interface MenuOverlayInternalProps extends OverlayInteractiveProps {
  arrow?: boolean
  backdropOffset?: {
    top?: string
    left?: string
    bottom?: string
    right?: string
  }
}

const MenuOverlayInternal: React.FC<MenuOverlayInternalProps> = ({
  arrow = true,
  children,
  backdropOffset,
  ...overlayProps
}) => {
  const backdrop = {
    background: 'transparent',
    ...backdropOffset,
  }

  return (
    <Overlay
      backdrop={backdrop}
      experimentalModifiers={{ flip: { enabled: false } }}
      {...overlayProps}
    >
      {(props: OverlayChildrenProps) => (
        <OverlaySurface
          arrow={arrow}
          {...props}
          {...CustomizableMenuOverlayAttributes.surface}
        >
          {children}
        </OverlaySurface>
      )}
    </Overlay>
  )
}

export interface MenuOverlayProps extends ModalManagerProps {
  arrow?: boolean
  backdropOffset?: {
    top?: string
    left?: string
    bottom?: string
    right?: string
  }
}

export class MenuOverlay extends ModalManager<MenuOverlayProps> {
  constructor(props: MenuOverlayProps) {
    super(props)

    this.checkClickOrigin = this.checkClickOrigin.bind(this)
  }

  public open() {
    this.setState(prevState => ({ isOpen: !prevState.isOpen }))
    window.addEventListener('keydown', this.handleEscapePress)
    document.addEventListener('mousedown', this.checkClickOrigin)
  }

  public checkClickOrigin(event: MouseEvent) {
    if (
      this.portalRef.current &&
      this.portalRef.current.contains(event.target as Node)
    ) {
      return
    }

    if (this.triggerRef.current) {
      if (!this.triggerRef.current.contains(event.target as Node)) {
        this.close()
      }
    }
  }

  protected renderModal(content: string, props: ManagedModalProps) {
    return (
      <MenuOverlayInternal
        isOpen={this.state.isOpen}
        triggerRef={this.triggerRef}
        onClose={this.close}
        {...props}
      >
        {content}
      </MenuOverlayInternal>
    )
  }
}

export interface CustomizableMenuOverlayAttributes
  extends CustomizableAttributes {
  surface: ModalSurfaceStyleProps
}

export const CustomizableMenuOverlayAttributes: CustomizableMenuOverlayAttributes = {
  surface: {
    backgroundColor: palette.white,
    border: '1px solid',
    borderColor: palette.charcoal200,
    borderRadius: radii.medium,
    boxShadow: shadows[3],
    color: palette.charcoal900,
  },
}
