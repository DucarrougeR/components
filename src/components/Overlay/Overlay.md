
Overlay is the abstract building block of Popover.
