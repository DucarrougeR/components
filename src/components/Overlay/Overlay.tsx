import { Placement } from 'popper.js'
import * as React from 'react'
import { Popper, PopperArrowProps } from 'react-popper'
import { ModalBackdrop, ModalContext } from '../Modal'
import { ModalPortal } from '../Modal/ModalPortal'

export interface OverlayContentProps {
  ref: React.Ref<HTMLElement>
  style: React.CSSProperties
  /**
   * Properties to be applied to the arrow container. These originate from the
   * underlying react-popper library and are used to position and style the
   * arrow.
   */
  arrowProps: PopperArrowProps
  /**
   * The placement (eg top, bottom, left-end, etc) of the arrow. Useful for
   * applying conditional styles to the arrow container.
   */
  placement: Placement

  eventHandlers?: React.DOMAttributes<{}>
}

export interface OverlayInteractiveProps {
  children: React.ReactNode

  /**
   * When true, renders the Backdrop, Surface and it's contained content immediately.
   * @default false
   */
  isOpen: boolean
  /**
   * Specify a callback to be called each time this Modal is closed
   */
  onClose: () => void

  triggerRef: HTMLElement | null

  /**
   * Pins popper placement and prevents popper from moving on window resize.
   * @default false
   */
  pin?: boolean
  /**
   * Can be one of: top, bottom, left, right, auto, with the modifiers: start,
   * end. This value comes directly from popperjs. See
   * https://popper.js.org/popper-documentation.html#Popper.placements for more
   * info.
   * @default bottom
   */
  placement?: Placement
}

export interface OverlayProps extends OverlayInteractiveProps {
  /**
   * Expects a callback that returns the actual overlay content to render. The
   * callback receives the following props:
   *
   * **arrowProps**: properties used to correctly position the arrow on a
   * content surface **placement**: the location of the arrow.
   *
   * See OverlaySurface.tsx for an example of how to use these properties.
   */
  children: (props: OverlayContentProps) => React.ReactNode
  /**
   * Optional backdrop styles to merge with the Backdrop implementation. These
   * must be a CSSProperty compatible key / value paired object. For example
   * {backgroundColor: 'pink'}.
   */
  backdropStyles?: React.CSSProperties
}

/**
 * Overlay is a low-level component that can be used to build Popovers,
 * Tooltips, Modals, etc. It handles coordinating a trigger component, like a
 * Button or Link, with the rendering of an "overlay" or "content surface" that
 * appears above other content in the page.
 *
 * Under the hood, Overlay is powered by [Popper.js](https://popper.js.org/) and
 * its corresponding [React library,
 * react-popper](https://github.com/FezVrasta/react-popper).
 */

export class Overlay extends React.Component<OverlayProps> {
  private portalRef: React.RefObject<HTMLElement>

  constructor(props: OverlayProps) {
    super(props)
    this.portalRef = React.createRef()
  }

  public componentDidMount() {
    window.addEventListener('keydown', this.handleEscapePress)
  }

  public componentWillUnmount() {
    window.removeEventListener('keydown', this.handleEscapePress)
  }

  public render() {
    const surface = (
      <ModalPortal ref={this.portalRef}>
        <ModalBackdrop
          onClick={this.props.onClose}
          style={this.props.backdropStyles}
        />
        <Popper
          positionFixed
          placement={this.props.placement}
          modifiers={{ flip: { enabled: this.props.pin ? false : true } }}
          referenceElement={
            this.props.triggerRef ? this.props.triggerRef : undefined
          }
        >
          {({ ref, style, arrowProps, placement }) =>
            this.props.children({
              arrowProps,
              placement,
              ref,
              style,
            })
          }
        </Popper>
      </ModalPortal>
    )

    return (
      <ModalContext.Provider value={{ closeModal: this.props.onClose }}>
        {this.props.isOpen && surface}
      </ModalContext.Provider>
    )
  }

  private handleEscapePress = (event: KeyboardEvent) => {
    if (event.key !== 'Escape') return
    if (!event.target) return
    if (
      !this.portalRef.current ||
      !this.portalRef.current.contains(event.target as Node)
    ) {
      return
    }

    this.props.onClose()
  }
}
