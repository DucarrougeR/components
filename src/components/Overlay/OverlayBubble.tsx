import { Placement } from 'popper.js'
import * as React from 'react'
import { PopperArrowProps } from 'react-popper'
import { fadeIn, styled, Theme } from '../../style'
import { Box, BoxProps } from '../Box'

export interface OverlayBubbleArrowProps {
  backgroundColor: string
  border: React.ReactText
  borderColor: React.ReactText
}

const OverlayBubbleArrow = styled.div<OverlayBubbleArrowProps>`
  position: absolute;

  &::before {
    content: '';
    margin: auto;
    display: block;
    width: 0.5rem;
    height: 0.5rem;
    background: ${props => props.backgroundColor};
    border-bottom: ${props => props.border} ${props => props.borderColor};
    border-right: ${props => props.border} ${props => props.borderColor};
  }

  &[data-placement*='top'] {
    bottom: 0.25rem;
    margin: 0 1rem;
    &::before {
      transform: rotate(45deg);
    }
  }

  &[data-placement*='right'] {
    left: 0.25rem;
    margin: 1rem 0;
    &::before {
      transform: rotate(135deg);
    }
  }

  &[data-placement*='bottom'] {
    top: 0.25rem;
    margin: 0 1rem;
    &::before {
      transform: rotate(225deg);
    }
  }

  &[data-placement*='left'] {
    right: 0.25rem;
    margin: 1rem 0;
    &::before {
      transform: rotate(315deg);
    }
  }
`

const OverlayBubbleContainer = styled<BoxProps<HTMLDivElement>>(Box)`
  animation: ${fadeIn} 0.2s linear;
`

export interface OverlayBubbleStyleProps extends OverlayBubbleArrowProps {
  borderColor: string
  borderRadius: string
  boxShadow: string
  color: string
  padding: keyof Theme['space']
}

export interface OverlayBubbleProps extends OverlayBubbleStyleProps {
  arrowProps: PopperArrowProps
  placement: Placement
}

export const OverlayBubble: React.SFC<OverlayBubbleProps> = ({ ...props }) => (
  <OverlayBubbleContainer
    p={props.padding}
    m="small"
    bg={props.backgroundColor}
    borderRadius={props.borderRadius}
    border={props.border}
    borderColor={props.borderColor}
    boxShadow={props.boxShadow}
    color={props.color}
  >
    {props.children}
    <OverlayBubbleArrow
      backgroundColor={props.backgroundColor}
      border={props.border}
      borderColor={props.borderColor}
      innerRef={props.arrowProps.ref}
      style={props.arrowProps.style}
      data-placement={props.placement}
    />
  </OverlayBubbleContainer>
)
