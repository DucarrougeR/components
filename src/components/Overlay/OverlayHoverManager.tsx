import { Placement } from 'popper.js'
import React, { useRef, useState } from 'react'
import { PopperProps } from 'react-popper'

export interface ManagedHoverOverlayProps {
  canClose: OverlayHoverManagerProps['canClose']
  close: () => void
  isOpen: boolean
  setSurfaceRef: NonNullable<PopperProps['innerRef']>
  triggerRef: React.RefObject<HTMLElement>
  onMouseOut: (event: React.MouseEvent) => void
}

export interface OverlayHoverManagerProps {
  __initializeOpenForLensTests?: boolean
  /**
   * Render Prop to render the controlled hover popper.
   * @required
   */
  children: (props: ManagedHoverOverlayProps) => React.ReactNode
  /**
   * Specify a callback to be called before trying to close the Modal. This allows for
   * use-cases where the user might lose work (think common "Save before closing warning" type flow)
   * Specify a callback to be called each time this Modal is closed
   */
  canClose?: () => boolean
  /**
   * Can be one of: top, bottom, left, right, auto, with the modifiers: start,
   * end. This value comes directly from popperjs. See
   * https://popper.js.org/popper-documentation.html#Popper.placements for more
   * info.
   * @default bottom
   */
  placement?: Placement
  /**
   * Component to wrap. The ModalHoverManager HOC will listen for mouse events on this
   * component, maintain the state of isOpen accordingly, and pass that state into
   * the modal renderProp.
   */
  wrappedComponent: (
    eventsHandlers: {
      onBlur: () => void
      onFocus: () => void
      onMouseOut: (event: React.MouseEvent) => void
      onMouseOver: () => void
    },
    ref: React.RefObject<HTMLElement>
  ) => React.ReactNode
}

export const OverlayHoverManager: React.FC<OverlayHoverManagerProps> = ({
  __initializeOpenForLensTests,
  canClose,
  ...props
}) => {
  const [isOpen, setIsOpen] = useState(__initializeOpenForLensTests || false)
  const surfaceRef = useRef<HTMLElement | null>(null)
  const triggerRef = useRef<HTMLElement>(null)

  const handleOpen = () => {
    setIsOpen(true)
  }

  const handleClose = () => {
    if (canClose && !canClose()) return
    setIsOpen(false)
  }

  const handleMouseOut = (event: React.MouseEvent) => {
    if (!isOpen) return

    const related = event.relatedTarget

    if (
      triggerRef.current &&
      (triggerRef.current === related ||
        triggerRef.current.contains(related as Node))
    ) {
      return
    }

    if (
      surfaceRef.current &&
      (surfaceRef.current === related ||
        surfaceRef.current.contains(related as Node))
    ) {
      return
    }

    handleClose()
  }

  const { children, wrappedComponent, ...otherProps } = props

  const eventHandlers = {
    onBlur: handleClose,
    onFocus: handleOpen,
    onMouseOut: handleMouseOut,
    onMouseOver: handleOpen,
  }

  const managedHoverOverlayProps = {
    ...otherProps,
    canClose,
    close: handleClose,
    isOpen,
    onMouseOut: handleMouseOut,
    setSurfaceRef: (ref: HTMLElement | null) => {
      surfaceRef.current = ref
    },
    triggerRef,
  }

  return (
    <>
      {children(managedHoverOverlayProps)}
      {wrappedComponent(eventHandlers, triggerRef)}
    </>
  )
}
