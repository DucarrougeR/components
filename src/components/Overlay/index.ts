export * from './Overlay'
export * from './OverlaySurface'
export * from './OverlayHover'
export * from './Popover'
export * from './PopoverContent'
export * from './RichTooltip'
export * from './RichTooltipContent'
export * from './Tooltip'
