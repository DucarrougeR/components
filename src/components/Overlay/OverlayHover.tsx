import * as React from 'react'
import { Popper } from 'react-popper'
import { CustomizableModalAttributes } from '../Modal'
import { OverlayProps } from './Overlay'

export interface OverlayHoverProps extends OverlayProps {
  setSurfaceRef?: (ref: null | HTMLElement) => void
  onMouseOut?: (event: React.MouseEvent) => void
}

export const OverlayHover: React.SFC<OverlayHoverProps> = ({
  children,
  isOpen,
  ...props
}) =>
  isOpen ? (
    <Popper
      positionFixed
      innerRef={props.setSurfaceRef}
      placement={props.placement}
      referenceElement={props.triggerRef ? props.triggerRef : undefined}
    >
      {({ ref, style, arrowProps, placement }) =>
        children({
          arrowProps,
          eventHandlers: { onMouseOut: props.onMouseOut },
          placement,
          ref,
          style: {
            ...style,
            zIndex: CustomizableModalAttributes.zIndex,
          },
        })
      }
    </Popper>
  ) : null
