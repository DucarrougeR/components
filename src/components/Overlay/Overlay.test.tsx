import { ReactWrapper } from 'enzyme'
import 'jest-styled-components'
import * as React from 'react'
import { mountWithTheme } from '../../../test/utils/create_with_theme'
import { assertSnapshot } from '../../../test/utils/snapshot'
import { Overlay, OverlayEvent } from './Overlay'

const SimpleContent = () => <div>Simple Overlay</div>
const simpleContentFactory = () => <SimpleContent />

const SimpleOverlay: React.SFC<{ trigger?: OverlayEvent }> = ({ ...props }) => (
  <Overlay overlayContentFactory={simpleContentFactory} trigger={props.trigger}>
    <button>Trigger</button>
  </Overlay>
)

const returnTriggerAndOverlay = (trigger?: OverlayEvent) => {
  const overlay = mountWithTheme(<SimpleOverlay trigger={trigger} />)
  return [overlay, overlay.find('button')]
}

const assertOverlayOpen = (overlay: ReactWrapper, open: boolean) => {
  expect(overlay.contains(<SimpleContent />)).toEqual(open)
}

describe('Overlay', () => {
  test('Generates a simple Overlay', () => {
    assertSnapshot(<SimpleOverlay />)
  })

  describe('trigger: hover', () => {
    let overlay: ReactWrapper
    let trigger: ReactWrapper
    beforeEach(() => ([overlay, trigger] = returnTriggerAndOverlay()))
    afterEach(() => overlay.unmount())

    test('opens & closes the overlay on hover', () => {
      assertOverlayOpen(overlay, false)
      trigger.simulate('mouseover')
      assertOverlayOpen(overlay, true)
      trigger.simulate('mouseout', {
        currentTarget: {},
        relatedTarget: document.body,
      })
      assertOverlayOpen(overlay, false)
    })

    test('mouseout the SimpleOverlayContent closes on hover', () => {
      assertOverlayOpen(overlay, false)
      trigger.simulate('mouseover')
      assertOverlayOpen(overlay, true)
      overlay.find(SimpleContent).simulate('mouseout', {
        currentTarget: {},
        relatedTarget: document.body,
      })
      assertOverlayOpen(overlay, false)
    })

    test('overlay opens & closes on focus', () => {
      assertOverlayOpen(overlay, false)
      trigger.simulate('focus')
      assertOverlayOpen(overlay, true)
      trigger.simulate('blur')
      assertOverlayOpen(overlay, false)
    })
  })

  describe('trigger: click', () => {
    let overlay: ReactWrapper
    let trigger: ReactWrapper
    beforeEach(() => ([overlay, trigger] = returnTriggerAndOverlay('click')))
    afterEach(() => overlay.unmount())

    test('Trigger click opens and closes the overlay', () => {
      assertOverlayOpen(overlay, false)
      trigger.simulate('click')
      assertOverlayOpen(overlay, true)
      trigger.simulate('click')
      assertOverlayOpen(overlay, false)
    })

    test('Trigger click renders a backdrop, clicking backdrop closes it', () => {
      assertOverlayOpen(overlay, false)
      trigger.simulate('click')
      const backdrop = overlay.find({ position: 'fixed' }).first()
      expect(backdrop.exists()).toEqual(true)
      backdrop.simulate('click')
      assertOverlayOpen(overlay, false)
    })
  })
})
