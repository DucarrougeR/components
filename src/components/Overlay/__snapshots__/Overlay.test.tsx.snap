// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Overlay Generates a simple Overlay 1`] = `null`;
